File: Fig02.R
## We successfully run the script in November 2024. Re 
# realized an issue while using the data through gtrendsR::gtrends. Thus, please use the more robust python script 

## Google Trends, Fig 2
library(gtrendsR)
library(ggplot2)
library(dplyr)
library(tidyr)
library(lubridate)
library(anytime)
#setwd("~/Documents/Research/Yardstick/BER/github/BER/")

keywords.all <- c("MPK", "Ministerpräsidentenkonferenz",
 "corona mpk", "beschlüsse mpk", 
 "ergebnisse ministerpräsidentenkonferenz")

mpk <- gtrends(keyword = keywords.all, 
               geo = "DE",
               time = "2020-03-01 2021-09-21",
               onlyInterest = TRUE, 
               tz = -120)

mpk.time <- mpk$interest_over_time 
## MPK-Tagungen: https://www.vdek.com/politik/corona-sars-cov-2-politik-archiv.html

mpk.dates <- as.Date(c("2020-03-12", "2020-03-16",
 "2020-03-22", "2020-04-01", "2020-04-15", "2020-04-30",
 "2020-05-06", "2020-05-26",  "2020-06-17", "2020-07-16", "2020-08-27", "2020-09-29","2020-10-14", "2020-10-28", "2020-11-16", "2020-11-25", "2020-12-13", "2021-01-05", "2021-01-19", "2021-02-01", "2021-02-10", "2021-03-03", "2021-03-19", "2021-03-22", "2021-04-26", "2021-05-27", "2021-08-10"))

ggplot(data = mpk.time,
       aes(x = as.Date(date), y = as.numeric(hits), 
           group = keyword, colour = keyword))+
  geom_line() +
  geom_vline(xintercept = as.numeric(mpk.dates), 
             col = "red", linetype = "dashed")

mpk.time.group <- mpk.time %>% group_by(date) %>% 
  reframe(hit.sum = sum(hits, na.rm = T)) %>% 
  mutate(hit.norm = hit.sum / max(hit.sum, na.rm = T) )

ggplot(data = mpk.time.group, aes(x = as.Date(date), y = hit.norm)) +
  geom_line(col = "purple") +
  geom_vline(xintercept = as.numeric(mpk.dates), 
             col = "forestgreen", linetype = "dotted", linewidth = 1.2) +
  scale_y_continuous(labels = scales::percent) +
  scale_x_date(date_breaks = "2 months", date_labels = "%b %Y") +
  labs(x = "", y = "Relative Interest") +
  theme_gray(base_size = 20)



rects <- data.frame(
   xmin = mpk.dates - 3,
   xmax = mpk.dates + 1,
   ymin = -Inf,
   ymax = Inf
 ) %>% 
   mutate(startdate = as.Date(xmin),
          enddate =   as.Date(xmax))

 mpkdates.df <- mpk.dates %>% as.data.frame(.) %>% setNames(c("date"))

# geom_rect(data = rects, aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), 
#fill = "green", alpha = 0.2, inherit.aes = FALSE)

all_dates <- seq.Date(from = min(mpk.df$mpk.time), to = max(mpk.df$mpk.time), by = "day")
 
mpkday.df <- data.frame(date = all_dates) %>%
   left_join(x = ., y = mpk.time, by = "date") %>%
   fill(hits, .direction = "down") 
 
 
ggplot(data = mpk.time.group, aes(x = as.Date(date), y = hit.norm)) +
  geom_line(col = "purple") +
  geom_point(data = mpkdates.df, aes(x = date, y = 0), 
             colour = "forestgreen", size = 4) +
  scale_y_continuous(labels = scales::percent) +
  scale_x_date(date_breaks = "2 months", date_labels = "%b %Y") +
  labs(x = "", y = "Relative Interest") +
  theme_gray(base_size = 20)
ggsave("~/Documents/Research/Yardstick/BER/mpk.pdf", 
       width = 16, height = 9)

ggplot(data = mpkday.df, 
       aes(x = as.Date(date), y = hits/100)) +
  geom_line(col = "purple") +
  geom_point(data = mpkdates.df, aes(x = date, y = 0), 
             colour = "forestgreen", size = 4) +
  scale_y_continuous(labels = scales::percent) +
  scale_x_date(date_breaks = "2 months", date_labels = "%b %Y") +
  labs(x = "", y = "Relative Interest") +
  theme_gray(base_size = 20)
ggsave("~/Documents/Research/Yardstick/BER/mpk.pdf", 
       width = 16, height = 9)

######################################################################
ggplot(data = mpk.time.group, aes(x = as.Date(date), y = hit.norm)) +
  geom_line(col = "purple") +
  geom_rect(data = rects, aes(xmin = startdate, xmax = enddate, 
                              ymin = -Inf, ymax = Inf), 
            fill = "green", alpha = 0.2, inherit.aes = FALSE) +
    scale_y_continuous(labels = scales::percent) +
  scale_x_date(date_breaks = "2 months", date_labels = "%b %Y") +
  labs(x = "", y = "Relative Interest") +
  theme_gray(base_size = 20)




#########################################################################
### new plot with python data:
ggplot(data = mpk.df, 
       aes(x = mpk.time, y = hits)) +
  geom_line() +
  scale_x_date(date_breaks = "2 months", date_labels = "%b %Y") +
  geom_point(data = mpkdates.df, aes(x = date, y = 0), 
             colour = "forestgreen", size = 4) 


## week-wise:
all_dates <- seq.Date(from = min(mpk.df$mpk.time), to = max(mpk.df$mpk.time), by = "day")

mpkday.df <- data.frame(mpk.time = all_dates) %>%
  left_join(x = ., y = mpk.df, by = "mpk.time") %>%
  fill(hits, .direction = "down")


ggplot(data = mpkday.df, 
       aes(x = mpk.time, y = hits/100)) +
  geom_rect(aes(xmin=as.Date("2020-03-02"), xmax=as.Date("2020-05-17"), ymin=-Inf, ymax=Inf), fill='gray97', alpha=.1, col = "gray35") +
  geom_rect(aes(xmin=as.Date("2020-09-28"), xmax=as.Date("2021-02-28"), ymin=-Inf, ymax=Inf), fill='gray97', alpha=.1, col = "gray35") +
  geom_rect(aes(xmin=as.Date("2021-03-01"), xmax=as.Date("2021-06-13"), ymin=-Inf, ymax=Inf), fill='gray97', alpha=.1, col = "gray35") +
  geom_line(col = "purple", linewidth = 1.4) +
  scale_x_date(date_breaks = "2 months", date_labels = "%b %Y") +
  scale_y_continuous(labels = scales::percent) +
  geom_point(data = mpkdates.df, aes(x = date, y = 0.0), 
             colour = "forestgreen", size = 5) +
  theme_gray(base_size = 22) +
  labs(x = "", y = "Relative Interest") 

#
ggsave("~/Documents/Research/Yardstick/BER/mpk2.pdf", 
       width = 16, height = 9)

File: Fig02.py
## Figure 2
from pytrends.request import TrendReq
import pandas as pd
import time
from datetime import date
import matplotlib.pyplot as plt

# --- 2. Define parameters ----------------------------------------------------
keywords_all = ["MPK", "Ministerpräsidentenkonferenz", "corona mpk", "beschlüsse mpk", "ergebnisse ministerpräsidentenkonferenz"]  # example list of keywords
geo_region = "DE"                # Germany
time_range = "2020-03-01 2021-09-21"  # same as in your R snippet

# --- 3. Connect to Google Trends --------------------------------------------
# hl = interface language; tz = timezone offset in minutes
pytrends = TrendReq(hl="de-DE", tz=120)

# --- 4. Download interest over time -----------------------------------------
# pytrends allows multiple keywords (max 5 per request)
pytrends.build_payload(kw_list=keywords_all,
                       timeframe=time_range,
                       geo=geo_region)

data = pytrends.interest_over_time()
data["hits"] = data.iloc[:, 0:5].sum(axis=1)


mpk_dates = [
    date(2020, 3, 12), date(2020, 3, 16), date(2020, 3, 22),
    date(2020, 4, 1),  date(2020, 4, 15), date(2020, 4, 30),
    date(2020, 5, 6),  date(2020, 5, 26),
    date(2020, 6, 17),
    date(2020, 7, 16),
    date(2020, 8, 27),
    date(2020, 9, 29),
    # date(2020, 10, 7),  # Telefonschalte
    date(2020, 10, 14), date(2020, 10, 28),
    date(2020, 11, 16), date(2020, 11, 25),
    # date(2020, 12, 2),
    date(2020, 12, 13),
    date(2021, 1, 5),  date(2021, 1, 19),
    date(2021, 2, 1),  date(2021, 2, 10),
    date(2021, 3, 3),  date(2021, 3, 19), date(2021, 3, 22),
    date(2021, 4, 26),
    date(2021, 5, 27),
    date(2021, 8, 10)
]

plt.figure(figsize=(10, 5))
plt.plot(data.index, data["hits"], color="steelblue", linewidth=2)

# --- 2. Aesthetics --------------------------------------
plt.title("Google Trends — Combined Interest (Hits) in Germany, 2020–2021", fontsize=13)
plt.xlabel("Date")
plt.ylabel("Search interest (sum of keywords)")
plt.grid(True, linestyle="--", alpha=0.6)

# vertical lines for MPK dates ----------
for d in mpk_dates:
    plt.axvline(d, color="red", linestyle="--", linewidth=1, alpha=0.6)

plt.tight_layout()
plt.savefig("Fig02Python.pdf", dpi = 300)
plt.show()

